lib/externalclient/huobi/huobi_client.go

```
@@ -0,0 +1,200 @@
+package huobi
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+const (
+	//HuobiEndpoint is base on
+	HuobiEndpoint = "https://api.huobi.pro"
+)
+
+//Client represent a huobi client for
+//calling to huobi endpoint
+type Client struct {
+	APIKey    string
+	SecretKey string
+	sugar     *zap.SugaredLogger
+}
+
+//NewHuobiClient return a new HuobiClient instance
+func NewHuobiClient(apiKey, secretKey string, sugar *zap.SugaredLogger) *Client {
+	return &Client{
+		APIKey:    apiKey,
+		SecretKey: secretKey,
+		sugar:     sugar,
+	}
+}
+
+func (hc *Client) sign(msg string) (string, error) {
+	mac := hmac.New(sha256.New, []byte(hc.SecretKey))
+	if _, err := mac.Write([]byte(msg)); err != nil {
+		hc.sugar.Errorw("encode message error", "error", err.Error())
+		return "", err
+	}
+	result := base64.StdEncoding.EncodeToString(mac.Sum(nil))
+	return result, nil
+}
+
+func (hc *Client) fillRequest(req *http.Request, signNeeded bool) {
+	if req.Method == http.MethodPost || req.Method == http.MethodPut || req.Method == http.MethodDelete {
+		req.Header.Add("Content-Type", "application/json")
+	}
+	if signNeeded {
+		q := req.URL.Query()
+		method := req.Method
+		auth := q.Encode()
+		hostname := req.URL.Hostname()
+		path := req.URL.Path
+		payload := strings.Join([]string{method, hostname, path, auth}, "\n")
+		signature, err := hc.sign(payload)
+		if err != nil {
+			log.Fatal(err)
```

please return the error.